<a href="https://colab.research.google.com/github/wordsbyanmol/PUS2024_AArora/blob/main/HW3(Prison)_Arora.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>


August 2015  
Modified September 2017, October 2018

Author: FBB

** reproduce the result in **
http://www.mdrc.org/sites/default/files/What%20Strategies%20Work%20for%20the%20Hard%20FR.pdf

In [2]:
import os
import sys
import numpy as np
import pylab as pl

#%pylab inline

# NULL HYPOTHESIS: the % of former prisoners employed after release is the same or lower for candidates who participated in the program as for the control group, significance level p=0.05

In [3]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


** $H_0: P_0 - P_1 \geq$    0 **
    
** $H_a: P_0 - P_1 $< 0    **
    
    
** $\alpha$ = 0.05    **

** this is a TEST OF PROPORTIONS. we use the Binomial distribution since it is a yes/no (bernulli) test for each subject: the former inmate was or was not ever employed in a CEO transitional job (second row in the table above): **

** $P_0=0.035, P_1=0.701$**

In [4]:

# first thing: choose your significance threshold
alpha=0.05

# we like fractions better then percentages. as a rule of thumb, either use fractions or counts.
# 0 for control group, 1 for program group

P_0 = 3.5 * 0.01
P_1 = 70.1 * 0.01

if P_0 - P_1 >= 0:
    # we are done
    print ("the Null holds")
else:
    print ("we must assess the statistical significance")

#sample sizes

n_0 = 409
n_1 = 564

#lets get the counts by multiplying by the sample size

Nt_0 = P_0 * n_0
Nt_1 = P_1 * n_1


we must assess the statistical significance


# WHICH  TEST?

samples, categorical data

 ** TWO OPTIONS z test, or chi-square test.  **

# START WITH Z TEST

** the z test compares the stanrard deviation of the expected distribution and the observed result. it tells you literally how many standard deviations from the tail an observation is, under the _assumption of normality_ **


must define the sample standard deviation (standard error)

In [5]:
#define the sample proportion first
sp = (P_0 * n_0 + P_1 * n_1) / (n_1 + n_0)
print (sp)

0.4210472764645426


** standard deviation of the sampling distribution the distribution is Binomial, the binomial stdev is **

(see a proof here!: http://stats.stackexchange.com/questions/29641/standard-error-for-the-mean-of-a-sample-of-binomial-random-variables!):

$\sqrt{\frac{p(1 - p)}{n}}$

for 2 samples this becomes

$\sqrt{ \frac{ \hat{p}(1 - \hat{p})} {n0} + \frac{ \hat{p}(1 - \hat{p})} {n1} }$

cfr: page 138 of Statistics in a Nutshell, eq. 5.12 and here http://stattrek.com/hypothesis-test/difference-in-proportions.aspx?Tutorial=AP

** $\hat{p} =\frac{p_0  n_0 + p_1  n_1}{n_0+n_1}$ **


Note that in the online version of Statistics in a Nutshell, $\bar{x}$  should be  $\hat{p}$!!

In [6]:
# I am goonna create a little one line function to calculate the standard error
# and to calculate p

p = lambda p0, p1, n0, n1: (p0 * n0 + p1 * n1) / (n0 + n1)
#standard error
se = lambda p, n0, n1: np.sqrt(p * (1 - p) * (1.0 / n0 + 1.0 / n1))

** z score: how many standard deviation away from the population parameter is my statistic? **

 $z=\frac{P_1-P_0}{\sigma}$

In [7]:
zscore = lambda p0, p1, s : (p0 - p1) / s
z_2y = zscore(P_1, P_0, se(p(P_0, P_1, n_0, n_1), n_0, n_1))
print (z_2y)

20.769786540820895


note that using p0-p1 or p1-p0 at the numerator is equivalent because the standardizes normal value of z has mean 0 (see image below) so that we can use the absolute value of the z score, or equivalently look for $P[Z<z]$ if z is positive, and $P[Z>z]$ if z is negative.

* if $p<\alpha$ : reject H0 **

IMPORTANT!! note that this P in the bottom line of the table is not the p-value, but

p-value = 1-P

In [8]:
## p-value for employment after 2 years:
## since the largest number we read off the table for is (way) smaller
## than the value for our statistic
## our p-value will be smaller than it would be if calculated using
## (e.g.) .9998 (and in fact using 1.0000 which is the largest number
## in the table). Using 0.9998 is a **conservative** approach.

p_2y = 1 - 0.9984


def report_result(p,a):
    print ('is the p value ' +
           '{0:.2f} smaller than the critical value {1:.2f}?'.format(p,a))
    if p < a:
        print ("YES!")
    else:
        print ("NO!")

    print ('the Null hypothesis is {}'.format(\
                            'rejected' if p < a  else 'not rejected') )


report_result(p_2y, alpha)

is the p value 0.00 smaller than the critical value 0.05?
YES!
the Null hypothesis is rejected


# what if we used the values for where the former inmate was or was not "Convicted of a felony" (row 10) in the Recidivism (Years 1-3)?

 Null hypothesis?
 $H_0$?
 $H_a$?

 $P_0 = ??, P_1= ??$

Null Hypothesis: The percentage of prisoners convicted of a felony without a training programme (control group) is equal or lower than with training progamme (1-3 years).

Alternate Hypothesis: The percentage of prisoners convicted of a felony without a training programme (control group) is higher than the proportion of prisoners with training progamme to recidivism (1-3 years).

The significance level p=0.05 that means our p-value must be greater than 95 percent to reject the null hypothesis.

MATEHMATICAL FORMULA

Ho: P0 - P1 >= 0

Ha: P0 - P1 < 0

P0= 11.7 P1= 10.0

 look up data table and insert the appropriate values to get the appropriate result! you can use the functions I defined above, with different arguments.

In [9]:
# first thing: choose your significance threshold
alpha=0.05

# fractions
# 0 for control group, 1 for program group

P_0 = 11.7 * 0.01
P_1 = 10.0 * 0.01

if P_0 - P_1 >= 0:
    # we are done
    print ("the Null holds")
else:
    print ("we must assess the statistical significance")

#sample sizes

n_0 = 409
n_1 = 568

#lets get the counts by multiplying by the sample size

Nt_0 = P_0 * n_0
Nt_1 = P_1 * n_1

the Null holds


In this case, the code finds that P₀ - P₁ >= 0, meaning the null hypothesis holds. This suggests that there is no statistically significant difference in the proportion of prisoners convicted of a felony between the control group and the program group.

# Now lets do it with the $\chi^2$ test

this analysis can also be done with the $\chi^2$ test, and the  $\chi^2$ distribution,

see  flowcharts for how to choose a test and "Statistics In a Nutshell Chapter 4", or http://math.hws.edu/javamath/ryan/ChiSquare.html (if you are really just interested in the formula at face value)

** The chisq statistics tests the statistics calculated as : **

$\chi^2 = \sum_{i} \frac{(observation_i - expectation_i)^2}{expectation_i}$

 against a chi sq distribution.
 If we talk about sample fractions  that is

$\chi^2 = \sum_i \frac{(f_{i,observed} - f_{i,expectated})^2}{f_{i,expected}}$

 Where _i_ indicates the sum over _each cell_.
 turns out this quantity is distributed according to a chi square distribution, so if i get the $\chi^2$ statistics i can compare it to the full chisq distribution and see how far in the tail it is

 The trickiest part (but not that tricky) is to figure out how to construct the table of values. please see Statistics In a Nutshell Chapter 4, for our data for example: Thisis called a CONTINGENCY TABLE

|                 |     success         | failure|    |               
|-----------------|:-------------------:|:-------------------:|---------------------------|
| test sample     | number of successes in test    | number of failures in test    | number members of test sample |
| control sample  | number of successes in control | number of failures in control | number members of control sample|
|                 | total successes                |  total failures               | number of all members         |

|employed in subsadized job |     employed          | not employed     |                   
|---------------------------|:---------------------:|------------------|---------------------------|
| test sample               | $0.701*564$           | $0.299*564$      | 564                       |
| control sample            | $0.035*409$           | $0.965*4.09$     | 409                       |
|                           |                       |                  |                           |
| total                     | 409.679               |  562.912         | 973                       |

 for each cell the expected ratio is the product of the total of all rows and all columns, devided by the total

** FOR EACH CELL: **
**  expected_cell = $\frac{row~tot * col~tot}{total}$ **


In [10]:
!wget https://raw.githubusercontent.com/fedhere/PUS2024/refs/heads/main/HW3/evalChisq.py

--2024-09-26 15:10:28--  https://raw.githubusercontent.com/fedhere/PUS2024/refs/heads/main/HW3/evalChisq.py
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 185.199.111.133, 185.199.109.133, 185.199.110.133, ...
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|185.199.111.133|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 325 [text/plain]
Saving to: ‘evalChisq.py’

evalChisq.py        100%[===================>]     325  --.-KB/s    in 0s      

2024-09-26 15:10:28 (2.83 MB/s) - ‘evalChisq.py’ saved [325/325]



In [11]:
from evalChisq import evalChisq

In [12]:
Ntot = 973 # a + b + c + d = tot

sample_values = np.array([[0.701 * 564, 0.299 * 564], [0.0305 * 409, 0.965 * 409]])

print (evalChisq(sample_values))

436.22346257536003


This number must be compared to the chi sq distribution.
You must calculate the number of degrees of freedom forthis experiment.
Generally: DOF = Number of observations - number of Independent Variables

so here DOF = 1. Now you can look at the table below and draw conclusions about the rejection of the Null

432 is hella larger then 3.84

why am i mentioning 3.84?

how does the chi square statistic that you derived compare?

please  state what that means in terms of your Null hypothesis in a markdown cell below!

Interpretation:

A chi-square test statistic of 436.22346257536003 with 1 degree of freedom is a very large value. This suggests a strong association between the two variables  in the data.

You mention 3.84 as that is the statistically significant value based on the chi square table.

# !!!!!! TODO FOR YOU!
lets see what the chi sq statistics says about the conviction for fellonies (row 10)

Hypothesis Testing:

Null Hypothesis (H₀): There is no association between group membership (control or program) and the outcome (convicted of a felony or not). In other words, the proportions of felony convictions are the same in both groups.

Alternative Hypothesis (Hₐ): There is an association between group membership and the outcome. In other words, the proportions of felony convictions are different between the two groups.

A chi-square test of independence is appropriate for this scenario. It tests whether there is a significant association between two categorical variables (group membership and felony conviction).

The alpa is 0.05%


* fill in the contingency table below *

|convicted of a fellony     |     yes   | no        |                   
|---------------------------|-----------|-----------|----------------|
| test sample               |     56.8  511.2     |   568   |                |
| control sample            |     47.85 |361.15     |409                |
|                           |           |           |                |
| total                     |    104.65 |872.35         977  |                ||


convicted of a fellony     |     yes   | no        |     total              
|---------------------------|-----------|-----------|----------------|
| test sample               |     56.8  511.2     |   568   |                |
| control sample            |     47.85 |361.15     |409                |
|                           |           |           |                |
| total                     |    104.65 |872.35         977  |                |

continue...

In [13]:
# fractions
# 0 for control group, 1 for program group

P_0 = 11.7 * 0.01
P_1 = 10.0 * 0.01

print(P_0)
print(P_1)

0.11699999999999999
0.1


In [15]:
Ntot = 977 # a + b + c + d = tot

sample_values = np.array([[0.1 * 568, 0.9 * 568], [0.116 * 409, 0.894 * 409]])

print (evalChisq(sample_values))

0.5554784876474075


A chi-square statistic of 0.55 suggests that there is no strong association between group membership (control or program) and the outcome (convicted of a felony or not). This means that the observed differences in the proportions of felony convictions between the two groups are likely due to chance, and there is not enough evidence to conclude that the program has a significant impact on recidivism rates related to felony convictions.

Interpretation:

P-value: The p-value is 0.455, which is significantly greater than the typical alpha level of 0.05. This means we fail to reject the null hypothesis.

Null Hypothesis: There is no association between group membership (control or program) and the outcome (convicted of a felony or not).

Conclusion: Based on the data, there is not enough evidence to suggest that the program has a significant impact on the likelihood of former prisoners being convicted of a felony. The observed differences between the two groups could be due to chance.